File: AnalyzeTemplate.java
import java.io.File;
import java.io.IOException;
import java.util.List;

import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.common.PDRectangle;
import org.apache.pdfbox.pdmodel.interactive.annotation.PDAnnotation;
import org.apache.pdfbox.pdmodel.interactive.annotation.PDAnnotationWidget;
import org.apache.pdfbox.pdmodel.interactive.form.PDAcroForm;
import org.apache.pdfbox.pdmodel.interactive.form.PDField;
import org.apache.pdfbox.text.PDFTextStripper;
import org.apache.pdfbox.text.TextPosition;

public class AnalyzeTemplate {

    public static void main(String[] args) throws Exception {

        if (args.length != 1) {
            System.out.println(
                "Usage: java AnalyzeTemplate <pdf>"
            );
            System.exit(1);
        }

        File pdfFile = new File(args[0]);

        try (PDDocument document = Loader.loadPDF(pdfFile)) {

            System.out.println("=================================");
            System.out.println("PDF TEMPLATE ANALYSIS");
            System.out.println("=================================");

            // =================================================
            // LAYER 1 — TECHNICAL
            // =================================================

            System.out.println();
            System.out.println("=== LAYER 1: TECHNICAL ===");

            System.out.println("PDF Type: PDF");

            int pageCount =
                document.getNumberOfPages();

            System.out.println(
                "Pages: " + pageCount
            );

            PDAcroForm acroForm =
                document.getDocumentCatalog().getAcroForm();

            boolean hasAcroForm =
                acroForm != null;

            System.out.println(
                "AcroForm: " +
                (hasAcroForm ? "Yes" : "No")
            );

            if (hasAcroForm) {

                int fieldCount = 0;

                for (PDField field :
                     acroForm.getFieldTree()) {

                    fieldCount++;

                    System.out.println();
                    System.out.println(
                        "Field: " +
                        field.getFullyQualifiedName()
                    );

                    System.out.println(
                        "Type: " +
                        field.getFieldType()
                    );

                    List<PDAnnotationWidget> widgets =
                        field.getWidgets();

                    if (widgets != null) {

                        for (PDAnnotationWidget widget :
                             widgets) {

                            PDRectangle rect =
                                widget.getRectangle();

                            if (rect != null) {

                                System.out.println(
                                    "Position: " +
                                    "x=" +
                                    rect.getLowerLeftX() +
                                    ", y=" +
                                    rect.getLowerLeftY() +
                                    ", width=" +
                                    rect.getWidth() +
                                    ", height=" +
                                    rect.getHeight()
                                );
                            }
                        }
                    }
                }

                System.out.println();
                System.out.println(
                    "Field Count: " + fieldCount
                );
            }

            // =================================================
            // LAYER 2 — CONTEXT
            // =================================================

            System.out.println();
            System.out.println("=== LAYER 2: CONTEXT ===");

            for (
                int pageNumber = 1;
                pageNumber <= pageCount;
                pageNumber++
            ) {

                PDPage page =
                    document.getPage(pageNumber - 1);

                PDRectangle pageSize =
                    page.getMediaBox();

                System.out.println();
                System.out.println(
                    "--- PAGE " +
                    pageNumber +
                    " ---"
                );

                System.out.println(
                    "Width: " +
                    pageSize.getWidth()
                );

                System.out.println(
                    "Height: " +
                    pageSize.getHeight()
                );

                // ---------------------------------------------
                // Printed text + positions
                // ---------------------------------------------

                PDFTextStripper stripper =
                    new PDFTextStripper() {

                        @Override
                        protected void writeString(
                            String text,
                            List<TextPosition> positions
                        ) throws IOException {

                            for (TextPosition position :
                                 positions) {

                                String value =
                                    position.getUnicode();

                                if (
                                    value != null &&
                                    !value.trim().isEmpty()
                                ) {

                                    System.out.println(
                                        "Text: " +
                                        value +
                                        " | x=" +
                                        position.getXDirAdj() +
                                        " | y=" +
                                        position.getYDirAdj() +
                                        " | width=" +
                                        position.getWidthDirAdj() +
                                        " | height=" +
                                        position.getHeightDir()
                                    );
                                }
                            }
                        }
                    };

                stripper.setStartPage(pageNumber);
                stripper.setEndPage(pageNumber);

                stripper.getText(document);

                // ---------------------------------------------
                // Form widgets
                // ---------------------------------------------

                System.out.println(
                    "--- FORM WIDGETS ---"
                );

                List<PDAnnotation> annotations =
                    page.getAnnotations();

                for (PDAnnotation annotation :
                     annotations) {

                    if (
                        annotation instanceof
                        PDAnnotationWidget
                    ) {

                        PDAnnotationWidget widget =
                            (PDAnnotationWidget)
                            annotation;

                        PDRectangle rect =
                            widget.getRectangle();

                        if (rect == null) {
                            continue;
                        }

                        System.out.println(
                            "Widget: " +
                            "x=" +
                            rect.getLowerLeftX() +
                            " | y=" +
                            rect.getLowerLeftY() +
                            " | width=" +
                            rect.getWidth() +
                            " | height=" +
                            rect.getHeight()
                        );
                    }
                }
            }

            System.out.println();
            System.out.println(
                "================================="
            );
            System.out.println(
                "ANALYSIS COMPLETE"
            );
            System.out.println(
                "================================="
            );
        }
    }
}

File: analyze-template.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { PDFDocument } from "pdf-lib";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function analyzePdfBytes(bytes, filename = "template.pdf") {
    const pdfjs = await import(
        "pdfjs-dist/legacy/build/pdf.mjs"
    );

    const uint8Bytes = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);

    // =========================================================
    // LAYER 1 — TECHNICAL
    // =========================================================

    const pdfDoc = await PDFDocument.load(uint8Bytes, { ignoreEncryption: true });
    const form = pdfDoc.getForm();

    const technicalFields = form.getFields().map(field => ({
        name: field.getName(),
        type: field.constructor.name
    }));

    // =========================================================
    // LAYER 2 — CONTEXT
    // =========================================================

    const loadingTask = pdfjs.getDocument({
        data: uint8Bytes
    });

    const pdf = await loadingTask.promise;

    const contextPages = [];

    for (let pageNumber = 1; pageNumber <= pdf.numPages; pageNumber++) {

        const page = await pdf.getPage(pageNumber);

        const viewport = page.getViewport({
            scale: 1
        });

        // -----------------------------------------------------
        // Read all printed PDF text
        // -----------------------------------------------------

        const textContent = await page.getTextContent();

        const text = textContent.items
            .filter(item =>
                item.str &&
                item.str.trim()
            )
            .map(item => ({
                text: item.str.trim(),

                x: item.transform[4],

                y:
                    viewport.height -
                    item.transform[5],

                width: item.width,

                height: item.height
            }));

        // -----------------------------------------------------
        // Read PDF form widgets
        // -----------------------------------------------------

        const annotations =
            await page.getAnnotations({
                intent: "display"
            });

        const fields = annotations
            .filter(annotation =>
                annotation.subtype === "Widget" &&
                annotation.fieldName
            )
            .map(annotation => {

                const rect = annotation.rect;

                return {
                    name: annotation.fieldName,

                    type: annotation.fieldType,

                    x: rect[0],

                    y:
                        viewport.height -
                        rect[3],

                    width:
                        rect[2] -
                        rect[0],

                    height:
                        rect[3] -
                        rect[1]
                };
            });

        // -----------------------------------------------------
        // Store complete page context
        // -----------------------------------------------------

        contextPages.push({
            page: pageNumber,

            size: {
                width: viewport.width,
                height: viewport.height
            },

            text,

            fields
        });
    }

    // =========================================================
    // FINAL TWO-LAYER RESULT
    // =========================================================

    const cleanFilename = String(filename || "template.pdf").replace(/\\/g, "/").split("/").pop();

    return {
        template: {
            filename: cleanFilename,
            path: cleanFilename
        },

        technical: {
            pdfType: "PDF",

            pages: pdfDoc.getPageCount(),

            hasAcroForm:
                technicalFields.length > 0,

            fieldCount:
                technicalFields.length,

            fields:
                technicalFields
        },

        context: {
            pages:
                contextPages
        }
    };
}

export async function analyzePdf(pdfPath) {
    const bytes = new Uint8Array(fs.readFileSync(pdfPath));
    const result = await analyzePdfBytes(bytes, path.basename(pdfPath));
    result.template.path = pdfPath;
    return result;
}

// =============================================================
// COMMAND LINE
// =============================================================

const isDirectCli = process.argv[1] && path.resolve(process.argv[1]) === __filename;

if (isDirectCli) {
    const pdfPath = process.argv[2];

    if (!pdfPath) {
        console.error(
            "Usage: node analyze-template.js <pdf>"
        );

        process.exit(1);
    }

    analyzePdf(pdfPath)
        .then(result => {

            const outputPath = process.argv[3] || path.join(
                __dirname,
                "template-analysis.json"
            );

            fs.writeFileSync(
                outputPath,
                JSON.stringify(
                    result,
                    null,
                    2
                )
            );

            console.log(
                "Template analysis complete."
            );

            console.log(
                `Pages: ${result.technical.pages}`
            );

            console.log(
                `AcroForm: ${
                    result.technical.hasAcroForm
                        ? "Yes"
                        : "No"
                }`
            );

            console.log(
                `Fields: ${result.technical.fieldCount}`
            );

            console.log(
                `Output: ${outputPath}`
            );
        })
        .catch(error => {

            console.error(
                "Template analysis failed:"
            );

            console.error(error);

            process.exit(1);
        });
}